api/client.go

```diff
@@ -413,18 +447,110 @@ func NewClient(c *Config) (*Client, error) {
 		u.Path = ""
 	}
 
+	c.modifyLock.Unlock()
+
 	client := &Client{
 		addr:   u,
-		config: c,
+		config: clientConfig,
+	}
+
+	// determine how to get a token
+	// order of precedence could change
+	tokenSinkPath := ""
+	switch {
+	case os.Getenv(EnvVaultToken) != "":
+		client.token = os.Getenv(EnvVaultToken)
+	case client.config.TokenFileSinkPath != "":
+		tokenSinkPath = client.config.TokenFileSinkPath
+	case client.config.AgentAddress != "":
+		tokenSinkPath, err = client.getSinkPathFromAgent(client.config.AgentAddress)
+		if err != nil {
+			return nil, errwrap.Wrapf(fmt.Sprintf("failed to determine tokenSinkPath given the provided agent address %q {{err}}", client.config.AgentAddress), err)
+		}
+		client.config.TokenFileSinkPath = tokenSinkPath
 	}
 
-	if token := os.Getenv(EnvVaultToken); token != "" {
+	// set and poll token from file sink if it is available
+	if client.config.TokenFileSinkPath != "" {
+		client.useFileSinkForToken = true
+		token, err := readTokenFromFile(client.config.TokenFileSinkPath)
+		if err != nil {
+			return nil, errwrap.Wrapf(fmt.Sprintf("failed to read token from file %q {{err}}", client.config.TokenFileSinkPath), err)
+		}
 		client.token = token
+		// poll file for updates
+		client.pollFileForToken()
 	}
 
 	return client, nil
 }
 
+// contacts agent for a file sink path
+func (c *Client) getSinkPathFromAgent(agentAddress string) (string, error) {
+	uri, err := url.Parse(agentAddress)
+	if err != nil {
+		return "", err
+	}
+	if uri.Path != "" {
+		return "", errors.New("configured agent address url should not specify a path")
+	}
+	uri.Path = consts.AgentPathFileSinks
+	if c.config.AgentSinkName != "" {
+		v := url.Values{}
+		v.Set("sinkName", c.config.AgentSinkName)
+		uri.RawQuery = v.Encode()
+	}
+
+	response, err := c.config.HttpClient.Get(uri.String())
+	if response != nil {
+		defer response.Body.Close()
+	}
+
+	responseBodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("received error code %d from agent: %s", response.StatusCode, string(responseBodyBytes))
```

Comment:
Not sure if we should include the response body in our error.  Get a second opinion maybe?

Reply:
@briankassouf requesting a second opinion on bubbling up the response body from the agent in a failed request to the agent here. Is there a convention to follow?